jacob-grahn/platform-racing-2-server
# Add Contests Hub

I doubt everything will work on the first try... but one can hope!

For this to have half a prayer of working on the first try, [**these changes will need to be made to the db**](https://github.com/jacob-grahn/platform-racing-2-server/issues/343).

Now, for a rundown of what this pull introduces...

Admins can:
 - Add/edit existing contests
 - Activate/deactivate contests (deletion of old contests occurs if they're not active and were last updated more than a year ago)
 - Add/remove prizes for contests
 - Do anything mods, contest owners, and regular users can do

^ When invoked, all of the admin-specific permissions are logged in the admin action log.

Mods can:
 - View expanded winner data (who awarded the prize, the IP address of that person, what prizes were awarded, and awarder comments)
 - Award a prize for any contest (unthrottled)
 - Do anything regular users can do

Contest Owners can:
 - Award prizes for their contest(s) (throttled; maximum number of times per week set in `max_awards` column of contests)
 - Do anything regular users can do

Regular Users can:
 - View winners of a contest
 - Get linked to a contest's page
 - Learn more about how to host contests

I tried very very hard to limit the number of requests that are being made to the database (i.e. logging the prize names as a joined array in the `prizes_awarded` column in the `contest_winners` table and retrieving that in the same db query as the rest of the data).  There are some places where a bunch of queries couldn't really be avoided without risking out-of-date things somewhere else down the line (i.e. On view_winners.php, I couldn't log the winner name, so I needed to get the name and power from the ID).  If you see anything that could be reduced, be my guest in editing the code as you see fit.

And as always, if you have any questions, let me know! :)
